### PR TITLE
Java: Add models for Commons-Lang's RegExUtils class

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

File: java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+public class RegExUtilsTest {
+  String taint() { return "tainted"; }
+
+  void sink(Object o) {}
+
+  void test() throws Exception {
+    Pattern cleanPattern = Pattern.compile("clean");
+    Pattern taintedPattern = Pattern.compile(taint());
+
+    sink(RegExUtils.removeAll(taint(), cleanPattern)); // $hasTaintFlow=y
```
Comment (Marcono1234) [inline java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java:13]: `=y` could be omitted, see #5347

File: java/change-notes/2021-03-05-regex-utils.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* Added models for Apache Commons-Lang's `RegExUtils` class. This means that any query that tracks tainted data may return additional results in cases where a `RegExUtils` transformation is part of the path from source to sink.
```
Comment (Marcono1234) [inline java/change-notes/2021-03-05-regex-utils.md:2]: Really minor, but it looks like the project itself does not use a hyphen, see https://commons.apache.org/proper/commons-lang/

```suggestion
* Added models for Apache Commons Lang's `RegExUtils` class. This means that any query that tracks tainted data may return additional results in cases where a `RegExUtils` transformation is part of the path from source to sink.
```

File: java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java (hunk 1 above)
Comment (smowton) [reply java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java:13]: I'll do this once that PR lasts

**Review by 447768 — APPROVED** [review]

**Comment by aschackmull** [comment]
These steps seem like something that could plausibly also be sanitisers in the context of some queries (and depending on the exact pattern). Historically I don't think we've modelled these sort of steps as taint steps for that reason. @lcartey Looks like you requested these steps - any insights on this?

**Comment by lcartey** [comment]
> These steps seem like something that could plausibly also be sanitisers in the context of some queries (and depending on the exact pattern). Historically I don't think we've modelled these sort of steps as taint steps for that reason. @lcartey Looks like you requested these steps - any insights on this?

I would prefer to model these as taint steps, and add them as sanitizers for specific queries where that makes sense. I think this makes an appropriate trade-off between false positives and false negatives. For example, I don't think we should consider any regex approach as a sanitizer for SQL injection.

**Comment by aschackmull** [comment]
> I would prefer to model these as taint steps, and add them as sanitizers for specific queries where that makes sense. I think this makes an appropriate trade-off between false positives and false negatives. For example, I don't think we should consider any regex approach as a sanitizer for SQL injection.

SGTM. We should make sure to give similar treatment to the jdk equivalents + `String.replace` etc., which I believe currently aren't taint steps. I'm guessing that the main query that want things like that as sanitizers is XSS (correct me if I'm wrong and/or more queries should be considered, @lcartey )

**Comment by smowton** [comment]
@aschackmull what order do you want these things to happen?

File: java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java (hunk 1 above)
Comment (aschackmull) [reply java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java:13]: It's merged now.

Comment (smowton) [reply java/ql/test/library-tests/frameworks/apache-commons-lang3/RegExUtilsTest.java]: Done

**Review by aschackmull — APPROVED** [review]

**Pull request merged**